doc/Messaging.md

```diff
@@ -0,0 +1,140 @@
+
+Messages can also be created to be sent in the future:
+
+```
+insert into my_message(id, message, time_scheduled) values(1, 'hello world', :future_time)
+```
+
+`future_time` must be the unix time expressed in nanoseconds.
+
+# Receiving messages
+
+Processes can subscribe to receive messages by sending a `MessageStream` request to VTGate. If there are multiple subscribers, the messages will be delivered in a round-robin fashion. Note that this is not a broadcast; Each message will be sent to at most one subscriber.
+
+The format for messages is the same as a vitess `Result`. This means that standard database tools that understand query results can also be message recipients. Currently, there is no SQL format for subscribing to messages, but one will be provided soon.
+
+## Subsetting
+
+It's possible that you may want to subscribe to specific shards or groups of shards while requesting messages. This is useful for partitioning or load balancing. The `MessageStream` API allows you to specify these constraints. The request parameters are as follows:
+
+* `Name`: Name of the message table.
+* `Keyspace`: Keyspace where the message table is present.
+* `Shard`: For unsharded keyspaces, this is usually "0". However, an empty shard will also work. For sharded keyspaces, a specific shard name can be specified.
+* `KeyRange`: If the keyspace is sharded, streaming will be performed only from the shards that match the range. This must be an exact match.
+
+# Acknowledging messages
+
+A received (or processed) message can be acknowledged using the `MessageAck` API call. This call accepts the following parameters:
+
+* `Name`: Name of the message table.
+* `Keyspace`: Keypsace where the message table is present. This field can be empty if the table name is unique across all keyspaces.
+* `Ids`: The list of ids that need to be acked.
+
+Once a message is successfully acked, it will never be resent.
+
+## Exponential backoff
+
+A message that was successfully sent will wait for the specified ack wait time. If no ack is received by then, it will be resent. The next attempt will be 2x the previous wait, and this delay is doubled for every attempt.
+
+# Purging
+
+Messages that have been successfully acked will be deleted after their age exceeds the time period specified by `vt_purge_after`.
+
+# Advanced usage
+
+The `MessageAck` functionality is currently an API call and cannot be used inside a transaction. However, you can ack a message using a regular DML. It should look like this:
+
+```
+update my_message set time_acked = :time_acked, time_next = null where id in ::ids and time_acked is null
+```
+
+You can also view messages using regular `select` queries.
+
+# Known limitations
+
+The message feature is currently in alpha, and can be improved. Here is the list of possible limitations/improvements:
+
+* Flexible columns: Allow any number of application defined columns to be in the message table.
```

Comment:
This won't be too hard to add on (after I take care of a few unrelated high-priority items).